### ISSUE TITLE
ndarray.cc:(.text+0x81a): undefined reference to `void mxnet::ndarray::EvalClip<mshadow::gpu>(mshadow::TBlob const&, float const&, float const&, mshadow::TBlob*, mxnet::RunContext)'

### ISSUE THREAD

**Issue by cuimiaomiao** [dmlc/mxnet#2053]
g++ -DMSHADOW_FORCE_STREAM -Wall -O3 -I./ -I/gruntdata/app_data/shumiao/svn/Mine/mxnet/mshadow/ -I/gruntdata/app_data/shumiao/svn/Mine/mxnet/dmlc-core/include -fPIC -Iinclude -msse3 -funroll-loops -Wno-unused-parameter -Wno-unknown-pragmas -I/gruntdata/app_data/shumiao/env/cuda/include -DMSHADOW_USE_CBLAS=1 -DMSHADOW_USE_MKL=0 -DMSHADOW_RABIT_PS=0 -DMSHADOW_DIST_PS=0 -DMSDHADOW_USE_PASCAL=0 -DMXNET_USE_OPENCV=1 -I/home/miaomiao.cmm/env/opencv-2.4.10/include  -fopenmp -DMSHADOW_USE_CUDNN=1  -DMXNET_USE_NVRTC=0 -std=c++0x  -o bin/im2rec tools/im2rec.cc build/src/resource.o build/src/c_api/c_api.o build/src/c_api/c_api_error.o build/src/c_api/c_predict_api.o build/src/common/mxrtc.o build/src/engine/engine.o build/src/engine/naive_engine.o build/src/engine/threaded_engine.o build/src/engine/threaded_engine_perdevice.o build/src/engine/threaded_engine_pooled.o build/src/io/image_aug_default.o build/src/io/io.o build/src/io/iter_csv.o build/src/io/iter_image_recordio.o build/src/io/iter_mnist.o build/src/kvstore/kvstore.o build/src/ndarray/ndarray.o build/src/ndarray/ndarray_function.o build/src/operator/activation.o build/src/operator/batch_norm.o build/src/operator/block_grad.o build/src/operator/broadcast_reduce_op.o build/src/operator/cast.o build/src/operator/concat.o build/src/operator/convolution.o build/src/operator/crop.o build/src/operator/cross_device_copy.o build/src/operator/cudnn_batch_norm.o build/src/operator/deconvolution.o build/src/operator/dropout.o build/src/operator/elementwise_binary_op.o build/src/operator/elementwise_binary_scalar_op.o build/src/operator/elementwise_sum.o build/src/operator/elementwise_unary_op.o build/src/operator/embedding.o build/src/operator/fully_connected.o build/src/operator/identity_attach_KL_sparse_reg.o build/src/operator/l2_normalization.o build/src/operator/leaky_relu.o build/src/operator/loss_binary_op.o build/src/operator/lrn.o build/src/operator/matrix_op.o build/src/operator/native_op.o build/src/operator/ndarray_op.o build/src/operator/operator.o build/src/operator/operator_util.o build/src/operator/pooling.o build/src/operator/regression_output.o build/src/operator/reshape.o build/src/operator/roi_pooling.o build/src/operator/slice_channel.o build/src/operator/softmax_activation.o build/src/operator/softmax_output.o build/src/operator/swapaxis.o build/src/operator/upsampling.o build/src/optimizer/optimizer.o build/src/optimizer/sgd.o build/src/storage/storage.o build/src/symbol/graph_executor.o build/src/symbol/graph_memory_allocator.o build/src/symbol/static_graph.o build/src/symbol/symbol.o /gruntdata/app_data/shumiao/svn/Mine/mxnet/dmlc-core/libdmlc.a -pthread -lm -lcudart -lcublas -lcurand -L/gruntdata/app_data/shumiao/env/cuda/lib64 -L/gruntdata/app_data/shumiao/env/cuda/lib -lcblas -lrt -L/home/miaomiao.cmm/env/opencv-2.4.10/lib -lopencv_calib3d -lopencv_contrib -lopencv_core -lopencv_features2d -lopencv_flann -lopencv_gpu -lopencv_highgui -lopencv_imgproc -lopencv_legacy -lopencv_ml -lopencv_nonfree -lopencv_objdetect -lopencv_ocl -lopencv_photo -lopencv_stitching -lopencv_superres -lopencv_ts -lopencv_video -lopencv_videostab -lcudnn 
build/src/ndarray/ndarray.o: In function `std::_Function_handler<void ()(mxnet::RunContext, mxnet::engine::CallbackOnComplete), void mxnet::Engine::PushSync<mxnet::ClipOp(mxnet::NDArray const&, float const&, float const&, mxnet::NDArray*)::{lambda(mxnet::RunContext)#2}>(mxnet::ClipOp(mxnet::NDArray const&, float const&, float const&, mxnet::NDArray*)::{lambda(mxnet::RunContext)#2}, mxnet::Context, std::vector<mxnet::engine::Var*, std::allocator<mxnet::engine::Var> > const&, std::vector<mxnet::engine::Var*, std::allocator<mxnet::engine::Var> > const, mxnet::FnProperty, int)::{lambda(mxnet::RunContext, mxnet::engine::CallbackOnComplete)#1}>::_M_invoke(std::_Any_data const&, mxnet::RunContext, mxnet::engine::CallbackOnComplete)':
ndarray.cc:(.text+0x81a): undefined reference to `void mxnet::ndarray::EvalClip<mshadow::gpu>(mshadow::TBlob const&, float const&, float const&, mshadow::TBlob*, mxnet::RunContext)'
build/src/ndarray/ndarray.o: In function `std::_Function_handler<void ()(mxnet::RunContext, mxnet::engine::CallbackOnComplete), void mxnet::Engine::PushSync<mxnet::SetValueOp(float const&, mxnet::NDArray*)::{lambda(mxnet::RunContext)#2}>(mxnet::SetValueOp(float const&, mxnet::NDArray*)::{lambda(mxnet::RunContext)#2}, mxnet::Context, std::vector<mxnet::engine::Var*, std::allocator<mxnet::engine::Var> > const&, std::vector<mxnet::engine::Var*, std::allocator<mxnet::engine::Var> > const, mxnet::FnProperty, int)::{lambda(mxnet::RunContext, mxnet::engine::CallbackOnComplete)#1}>::_M_invoke(std::_Any_data const&, mxnet::RunContext, mxnet::engine::CallbackOnComplete)':
ndarray.cc:(.text+0x921): undefined reference to `void mxnet::ndarray::Eval<mshadow::gpu>(float const&, mshadow::TBlob*, mxnet::RunContext)'
build/src/ndarray/ndarray.o: In function `std::_Function_handler<void ()(mxnet::RunContext, mxnet::engine::CallbackOnComplete), void mxnet::Engine::PushSync<mxnet::CopyFromTo(mxnet::NDArray const&, mxnet::NDArray*, int)::{lambda(mxnet::RunContext)#2}>(mxnet::CopyFromTo(mxnet::NDArray const&, mxnet::NDArray*, int)::{lambda(mxnet::RunContext)#2}, mxnet::Context, std::vector<mxnet::engine::Var*, std::allocator<mxnet::engine::Var> > const&, std::vector<mxnet::engine::Var*, std::allocator<mxnet::engine::Var> > const, mxnet::FnProperty, int)::{lambda(mxnet::RunContext, mxnet::engine::CallbackOnComplete)#1}>::_M_invoke(std::_Any_data const&, mxnet::RunContext, mxnet::engine::CallbackOnComplete)':
ndarray.cc:(.text+0xa55): undefined reference to `void mxnet::ndarray::Copy<mshadow::cpu, mshadow::gpu>(mshadow::TBlob const&, mshadow::TBlob*, mxnet::Context, mxnet::Context, mxnet::RunContext)'
build/src/ndarray/ndarray.o: In function `std::_Function_handler<void ()(mxnet::RunContext, mxnet::engine::CallbackOnComplete), void mxnet::Engine::PushSync<mxnet::CopyFromTo(mxnet::NDArray const&, mxnet::NDArray*, int)::{lambda(mxnet::RunContext)#3}>(mxnet::CopyFromTo(mxnet::NDArray const&, mxnet::NDArray*, int)::{lambda(mxnet::RunContext)#3}, mxnet::Context, std::vector<mxnet::engine::Var*, std::allocator<mxnet::engine::Var> > const&, std::vector<mxnet::engine::Var*, std::allocator<mxnet::engine::Var> > const, mxnet::FnProperty, int)::{lambda(mxnet::RunContext, mxnet::engine::CallbackOnComplete)#1}>::_M_invoke(std::_Any_data const&, mxnet::RunContext, mxnet::engine::CallbackOnComplete)':
ndarray.cc:(.text+0xbc5): undefined reference to `void mxnet::ndarray::Copy<mshadow::gpu, mshadow::cpu>(mshadow::TBlob const&, mshadow::TBlob*, mxnet::Context, mxnet::Context, mxnet::RunContext)'
build/src/ndarray/ndarray.o: In function `std::_Function_handler<void ()(mxnet::RunContext, mxnet::engine::CallbackOnComplete), void mxnet::Engine::PushSync<mxnet::CopyFromTo(mxnet::NDArray const&, mxnet::NDArray*, int)::{lambda(mxnet::RunContext)#4}>(mxnet::CopyFromTo(mxnet::NDArray const&, mxnet::NDArray*, int)::{lambda(mxnet::RunContext)#4}, mxnet::Context, std::vector<mxnet::engine::Var*, std::allocator<mxnet::engine::Var> > const&, std::vector<mxnet::engine::Var*, std::allocator<mxnet::engine::Var> > const, mxnet::FnProperty, int)::{lambda(mxnet::RunContext, mxnet::engine::CallbackOnComplete)#1}>::_M_invoke(std::_Any_data const&, mxnet::RunContext, mxnet::engine::CallbackOnComplete)':
ndarray.cc:(.text+0xd35): undefined reference to `void mxnet::ndarray::Copy<mshadow::gpu, mshadow::gpu>(mshadow::TBlob const&, mshadow::TBlob*, mxnet::Context, mxnet::Context, mxnet::RunContext)'
build/src/ndarray/ndarray.o: In function `mxnet::NDArray::SyncCopyFromCPU(void const*, unsigned long) const':
ndarray.cc:(.text+0x1405): undefined reference to `void mxnet::ndarray::Copy<mshadow::cpu, mshadow::gpu>(mshadow::TBlob const&, mshadow::TBlob*, mxnet::Context, mxnet::Context, mxnet::RunContext)'
build/src/ndarray/ndarray.o: In function `mxnet::NDArray::SyncCopyToCPU(void*, unsigned long) const':
ndarray.cc:(.text+0x1b45): undefined reference to `void mxnet::ndarray::Copy<mshadow::gpu, mshadow::cpu>(mshadow::TBlob const&, mshadow::TBlob*, mxnet::Context, mxnet::Context, mxnet::RunContext)'
build/src/ndarray/ndarray.o: In function `std::_Function_handler<void ()(mxnet::RunContext, mxnet::engine::CallbackOnComplete), void mxnet::Engine::PushSync<mxnet::ElementwiseSum(std::vector<mxnet::NDArray, std::allocator<mxnet::NDArray> > const&, mxnet::NDArray*, int)::{lambda(mxnet::RunContext)#2}>(mxnet::ElementwiseSum(std::vector<mxnet::NDArray, std::allocator<mxnet::NDArray> > const&, mxnet::NDArray*, int)::{lambda(mxnet::RunContext)#2}, mxnet::Context, std::vector<mxnet::engine::Var*, std::allocator<mxnet::engine::Var> > const&, std::vector<mxnet::engine::Var*, std::allocator<mxnet::engine::Var> > const, mxnet::FnProperty, int)::{lambda(mxnet::RunContext, mxnet::engine::CallbackOnComplete)#1}>::_M_invoke(std::_Any_data const&, mxnet::RunContext, mxnet::engine::CallbackOnComplete)':
ndarray.cc:(.text+0x28c1): undefined reference to `void mxnet::ndarray::ElementwiseSum<mshadow::gpu>(std::vector<mshadow::TBlob, std::allocator<mshadow::TBlob> >, mshadow::TBlob*, mxnet::RunContext)'
build/src/ndarray/ndarray.o: In function `mxnet::Broadcast(mxnet::NDArray const&, int, int, mxnet::NDArray*)::{lambda(mxnet::RunContext)#2}::operator()(mxnet::RunContext) const':
ndarray.cc:(.text+0x3a67): undefined reference to `void mxnet::ndarray::EvalBroadcast<mshadow::gpu>(mshadow::TBlob const&, mshadow::TBlob*, int, mxnet::RunContext)'
build/src/ndarray/ndarray.o: In function `std::_Function_handler<void ()(mxnet::RunContext, mxnet::engine::CallbackOnComplete), void mxnet::Engine::PushSync<void mxnet::SampleOP<mxnet::ndarray::UniformDistribution>(float const&, float const&, mxnet::NDArray*)::{lambda(mxnet::RunContext)#2}>(void mxnet::SampleOP<mxnet::ndarray::UniformDistribution>(float const&, float const&, mxnet::NDArray*)::{lambda(mxnet::RunContext)#2}, mxnet::Context, std::vector<mxnet::engine::Var*, std::allocator<mxnet::engine::Var> > const&, std::vector<mxnet::engine::Var*, std::allocator<mxnet::engine::Var> > const, mxnet::FnProperty, int)::{lambda(mxnet::RunContext, mxnet::engine::CallbackOnComplete)#1}>::_M_invoke(std::_Any_data const&, mxnet::RunContext, mxnet::engine::CallbackOnComplete)':
ndarray.cc:(.text._ZNSt17_Function_handlerIFvN5mxnet10RunContextENS0_6engine18CallbackOnCompleteEEZNS0_6Engine8PushSyncIZNS0_8SampleOPINS0_7ndarray19UniformDistributionEEEvRKfSB_PNS0_7NDArrayEEUlS1_E0_EEvT_NS0_7ContextERKSt6vectorIPNS2_3VarESaISJ_EESN_NS0_10FnPropertyEiEUlS1_S3_E_E9_M_invokeERKSt9_Any_dataS1_S3_[_ZNSt17_Function_handlerIFvN5mxnet10RunContextENS0_6engine18CallbackOnCompleteEEZNS0_6Engine8PushSyncIZNS0_8SampleOPINS0_7ndarray19UniformDistributionEEEvRKfSB_PNS0_7NDArrayEEUlS1_E0_EEvT_NS0_7ContextERKSt6vectorIPNS2_3VarESaISJ_EESN_NS0_10FnPropertyEiEUlS1_S3_E_E9_M_invokeERKSt9_Any_dataS1_S3_]+0x59): undefined reference to `void mxnet::ndarray::EvalRandom<mshadow::gpu, mxnet::ndarray::UniformDistribution>(float const&, float const&, mxnet::Resource const&, mshadow::TBlob*, mxnet::RunContext)'
build/src/ndarray/ndarray.o: In function `std::_Function_handler<void ()(mxnet::RunContext, mxnet::engine::CallbackOnComplete), void mxnet::Engine::PushSync<void mxnet::SampleOP<mxnet::ndarray::GaussianDistribution>(float const&, float const&, mxnet::NDArray*)::{lambda(mxnet::RunContext)#2}>(void mxnet::SampleOP<mxnet::ndarray::GaussianDistribution>(float const&, float const&, mxnet::NDArray*)::{lambda(mxnet::RunContext)#2}, mxnet::Context, std::vector<mxnet::engine::Var*, std::allocator<mxnet::engine::Var> > const&, std::vector<mxnet::engine::Var*, std::allocator<mxnet::engine::Var> > const, mxnet::FnProperty, int)::{lambda(mxnet::RunContext, mxnet::engine::CallbackOnComplete)#1}>::_M_invoke(std::_Any_data const&, mxnet::RunContext, mxnet::engine::CallbackOnComplete)':
ndarray.cc:(.text._ZNSt17_Function_handlerIFvN5mxnet10RunContextENS0_6engine18CallbackOnCompleteEEZNS0_6Engine8PushSyncIZNS0_8SampleOPINS0_7ndarray20GaussianDistributionEEEvRKfSB_PNS0_7NDArrayEEUlS1_E0_EEvT_NS0_7ContextERKSt6vectorIPNS2_3VarESaISJ_EESN_NS0_10FnPropertyEiEUlS1_S3_E_E9_M_invokeERKSt9_Any_dataS1_S3_[_ZNSt17_Function_handlerIFvN5mxnet10RunContextENS0_6engine18CallbackOnCompleteEEZNS0_6Engine8PushSyncIZNS0_8SampleOPINS0_7ndarray20GaussianDistributionEEEvRKfSB_PNS0_7NDArrayEEUlS1_E0_EEvT_NS0_7ContextERKSt6vectorIPNS2_3VarESaISJ_EESN_NS0_10FnPropertyEiEUlS1_S3_E_E9_M_invokeERKSt9_Any_dataS1_S3_]+0x59): undefined reference to `void mxnet::ndarray::EvalRandom<mshadow::gpu, mxnet::ndarray::GaussianDistribution>(float const&, float const&, mxnet::Resource const&, mshadow::TBlob*, mxnet::RunContext)'
.........................

**Comment by cuimiaomiao** [comment]
My problem is like https://github.com/dmlc/mxnet/issues/1528 ,but I cannot use mkl instead of atlas. How can I do?

**Comment by tqchen** [comment]
you can try to use openblas instead

**Comment by muyinanhai** [comment]
Same error as above.(all of openblas atlas mkl blas output the same error)

System:
 Ubuntu 14.04
gcc version 4.8.4

config.mk:
USE_CUDA = 1

# add the path to CUDA library to link and compile flag
# if you have already add them to environment variable, leave it as NONE
USE_CUDA_PATH = /usr/local/cuda-7.5

UNAME_S := $(shell uname -s)
ifeq ($(UNAME_S), Darwin)
USE_BLAS = apple
else
USE_BLAS = openblas
endif

**Comment by cuimiaomiao** [comment]
Have tried all of them openblas atlas mkl blas, the same error....

**Comment by cuimiaomiao** [comment]
I soved this problem by myself.
Becase there was more  blank characters in the line of 'USE_CUDA = 1'....
It works well by repalcing 'USE_CUDA = 1 ' with 'USE_CUDA = 1'.

**Comment by lzhx171** [comment]
@cuimiaomiao Nice job. Thanks！

**Comment by lzhx171** [comment]
@cuimiaomiao  I was troubled by this problem nearly half a year!!!!!!!!!!!

**Comment by maxenceliu** [comment]
@cuimiaomiao Why??? Don't Understand!  just a blank?
